src/Authenticator/PasetoAuthenticator.php

```diff
@@ -0,0 +1,200 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         1.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authentication\Authenticator;
+
+use Authentication\Identifier\IdentifierInterface;
+use Exception;
+use ParagonIE\Paseto\JsonToken;
+use ParagonIE\Paseto\Keys\AsymmetricSecretKey;
+use ParagonIE\Paseto\Keys\SymmetricKey;
+use ParagonIE\Paseto\Parser;
+use ParagonIE\Paseto\Protocol\Version4;
+use ParagonIE\Paseto\ProtocolCollection;
+use ParagonIE\Paseto\ProtocolInterface;
+use ParagonIE\Paseto\Purpose;
+use Psr\Http\Message\ServerRequestInterface;
+use RuntimeException;
+
+/**
+ * PASETO Authenticator
+ *
+ * Authenticates an identity based on a PASETO token.
+ *
+ * @link https://github.com/paragonie/paseto
+ * @link https://github.com/paseto-standard/paseto-spec
+ */
+class PasetoAuthenticator extends TokenAuthenticator
+{
+    public const LOCAL = 'local';
+    public const PUBLIC = 'public';
+
+    /**
+     * @inheritDoc
+     */
+    protected $_defaultConfig = [
+        'header' => 'Authorization',
+        'queryParam' => 'token',
+        'tokenPrefix' => 'bearer',
+        'returnPayload' => true,
+        'version' => null,
+        'purpose' => null,
+        'secretKey' => null,
+        //'subjectKey' => IdentifierInterface::CREDENTIAL_JWT_SUBJECT,
+    ];
+
+    /**
+     * Payload data.
+     *
+     * @var object|null
+     */
+    protected $payload;
+
+    /**
+     * @inheritDoc
+     */
+    public function __construct(IdentifierInterface $identifier, array $config = [])
+    {
+        parent::__construct($identifier, $config);
+
+        if ($this->getConfig('version') === null) {
+            $this->setConfig('version', new Version4());
+        }
+        if (!$this->getConfig('version') instanceof ProtocolInterface) {
+            throw new RuntimeException(
+                'PASETO `version` config must be an instance of ' . ProtocolInterface::class
+            );
+        }
+
+        if (!in_array($this->getConfig('purpose'), [self::PUBLIC, self::LOCAL])) {
+            throw new RuntimeException('PASETO `purpose` config must one of: local or public');
+        }
+
+        if (empty($this->getConfig('secretKey'))) {
+            if (!class_exists(\Cake\Utility\Security::class)) {
+                throw new RuntimeException('PASETO `secretKey` config must be defined');
+            }
+            $this->setConfig('secretKey', \Cake\Utility\Security::getSalt());
```

Comment:
Will the application salt always be long enough? Are there length requirements in paseto?

Reply:
> This is where I wish the library provided a cli to easily generate those for developers.

We could provide that CLI tool :smile:

Reply:
>For Symmetric keys between 16 and 64 bytes. sodium_crypto_generichash will throw a SodiumException: unsupported key length exception otherwise, unfortunately this is not well documented here

That will work. If users get tripped up by the exceptions from libsodium we can always catch errors and rethrow with a better message.